src/zarr/core/buffer/core.py

```diff
@@ -457,7 +457,10 @@ def __repr__(self) -> str:
         return f"<NDBuffer shape={self.shape} dtype={self.dtype} {self._data!r}>"
 
     def all_equal(self, other: Any) -> bool:
-        return bool((self._data == other).all())
+        # use array_equal to obtain equal_nan=True functionality
+        data, other = np.broadcast_arrays(self._data, other)
```

Comment:
Does this create a copy of the data (or `other`)? If so it might be wise not the broadcast these, and just let `array_equal` error out?

Alternatively, maybe we should copy numpy behaviour and return `False` when comparing `NaN`s? Then I think we could use [`array_equiv`](https://numpy.org/doc/stable/reference/generated/numpy.array_equiv.html#numpy.array_equiv).

Reply:
`broadcast_to` always returns a [view](https://numpy.org/doc/stable/reference/generated/numpy.broadcast_to.html)
> A readonly view on the original array with the given shape. It is typically not contiguous. Furthermore, more than one element of a broadcasted array may refer to a single memory location.

> lternatively, maybe we should copy numpy behaviour and return False when comparing NaNs? 

I think the intent here is to not write all-fill-value chunks and it needs to handle a NaN fill value. I've added a unit test for this case.

Reply:
👍 makes sense. Since this is public API, I would be pro adding a new `equal_nan` keyword argument that defaults to `True`. This would let users change the behaviour, and also signals that the default beaviour is different to `np.array_equal`.

Reply:
added the kwarg